File: Sources/SlopDeskWorkspaceCore/Terminal/CopyReceipt.swift
// CopyReceipt — the "what just landed on the clipboard" summary behind every transient copy
// confirmation (the pane chip + the window-level chip), as a face over
// `slopdesk_terminal::copy_receipt`.
//
// A copy is the highest-frequency invisible action in a terminal, so the confirmation must answer the
// one real doubt — "did I get the whole thing?" — in a glance: a LINE count for a multi-line grab
// (where the selection may extend past the viewport), a CHAR count for a single-line grab (where
// truncation is the failure mode). Which number, how it is grouped, and the sentence it sits in are
// all the crate's; counting and wording crossed together because they are one answer, and a chip
// holding a count from one grab beside a sentence from the next is exactly what two doors would
// allow.

import CSlopDeskFFI
import Foundation
import SlopDeskWorkspaceModel

/// One clipboard-copy receipt: the counts of what was written, plus a monotonically-increasing `epoch`
/// so a rapid re-copy reads as a NEW receipt (the chip's dwell timer restarts and the label hard-cuts
/// to the new count — retarget, never a re-entrance animation).
/// `Hashable` over the same fields it is `Equatable` over — the AppKit chip keys its dwell on the
/// WHOLE receipt (a hand-off between the two owners can carry the same epoch), and an identity a view
/// compares has to be one it can also box.
public struct CopyReceipt: Hashable, Sendable {
    /// Grapheme-cluster count of the copied text — what a user would call "characters".
    public let charCount: Int
    /// Logical line count: newline-separated segments, with a single trailing newline NOT counted as an
    /// extra empty line (a shell line copy `"foo\n"` is one line, not two).
    public let lineCount: Int
    /// Bumped per copy by the owner (pane model / overlay coordinator) — identity for the dwell timer.
    public let epoch: Int
    /// The count half of the label: `"18 lines"` for a multi-line copy (the whole-block doubt),
    /// `"1,204 characters"` / `"1 character"` for a single line (the truncation doubt).
    public let detail: String
    /// The full label (`"Copied · 1,204 characters"`) — one string for accessibility + tests; the chip
    /// renders the two halves at separate weights (label quiet, count semibold).
    public let label: String

    /// How long the chip dwells before expiring — long enough to read a short count at a glance, short
    /// enough that the pane is ornament-free again before the next thought.
    ///
    /// It is a CONSTANT rather than per-receipt, unlike `ChipNotice.dwell`, because every copy says the
    /// same shape of thing; a notice may be offering an undo chord, which needs more reading time. It
    /// lives on the value rather than on either chip so the two renderers cannot hold the receipt for
    /// two different lengths of time. A DURATION and not a number of characters, so it is the one
    /// field here with nothing to decide and no reason to cross.
    public static let dwell: Duration = .seconds(1.5)

    public init(text: String, epoch: Int) {
        var bytes = Array(text.utf8)
        let blob = bytes.withUnsafeMutableBufferPointer { lent in
            wsAnswerBytes { out, cap in
                Int(slopdesk_copy_receipt(lent.baseAddress, lent.count, out, cap))
            }
        }
        let head = Int(SLOPDESK_COPY_RECEIPT_HEAD_BYTES)
        // The door never answers nothing — an empty copy still has a receipt — so a short delivery is
        // a layout disagreement, and zeroes beside empty words say so rather than half-reporting.
        let number = { (at: Int) -> Int in
            guard blob.count >= head else { return 0 }
            return (0..<4).reduce(0) { $0 << 8 | Int(blob[at + $1]) }
        }
        charCount = number(0)
        lineCount = number(4)
        let runs = wsRuns(blob.count >= head ? Array(blob.dropFirst(head)) : [], count: 2)
        detail = runs[0]
        label = runs[1]
        self.epoch = epoch
    }
}

File: rust/slopdesk-terminal/src/copy_receipt.rs
//! What just landed on the clipboard, and the one sentence that says so.
//!
//! A copy is the highest-frequency INVISIBLE action in a terminal, so the transient confirmation
//! has to answer the one real doubt — "did I get the whole thing?" — in a glance. Which number
//! answers it depends on the shape of the grab: a multi-line selection may extend past the
//! viewport, so its doubt is about LINES; a single-line grab is fully visible, so its doubt is
//! about truncation, which is CHARACTERS. Hence the ladder in [`detail`]: never "1 line", always
//! the more informative number.
//!
//! Counting and wording live together because they are one answer. Two chips draw this — the pane's
//! and the window's — and a receipt whose count came from here while its sentence came from a view
//! would be the two-renderer split the rail's git line already paid for once.

use unicode_segmentation::UnicodeSegmentation;

/// The two figures a receipt carries.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Default)]
pub struct Counts {
    /// Grapheme-CLUSTER count — what a user would call "characters", and what Swift's
    /// `String.count` answers, which is the number the label used to print.
    pub chars: usize,
    /// Logical lines: newline-separated segments, with a SINGLE trailing newline not counted as an
    /// extra empty one.
    pub lines: usize,
}

/// Counts what was copied.
///
/// The trailing-newline rule is the whole subtlety: a shell line copy arrives as `"foo\n"`, and
/// calling that two lines would report a phantom empty one on the most ordinary copy there is. Only
/// ONE trailing newline is absorbed — `"a\n\n"` really does end with a blank line, and a receipt
/// that hid it would be describing a different clipboard.
#[must_use]
pub fn counts(text: &str) -> Counts {
    let chars = text.graphemes(true).count();
    let mut segments: Vec<&str> = text.split('\n').collect();
    if segments.len() > 1 && segments.last().is_some_and(|last| last.is_empty()) {
        segments.pop();
    }
    Counts {
        chars,
        lines: segments.len(),
    }
}

/// The count half of the label: `"18 lines"` for a multi-line grab, `"1,204 characters"` /
/// `"1 character"` for a single line.
#[must_use]
pub fn detail(counts: Counts) -> String {
    if counts.lines > 1 {
        return format!("{} lines", grouped(counts.lines));
    }
    let noun = if counts.chars == 1 {
        "character"
    } else {
        "characters"
    };
    format!("{} {noun}", grouped(counts.chars))
}

/// The full sentence — `"Copied · 1,204 characters"`.
///
/// ⚠️ SENTENCE CASE, and the words are spelled out. Both went with the surface: the receipt moved
/// off the glass onto the floating family's paper capsule, whose voice is the system's neutral
/// semantics in sentence case, so the instrument caps register stayed with the glass it belonged
/// to. "CHARS" was an abbreviation that register needed to stay narrow; a proportional face at
/// reading size does not, and "characters" is what the number actually counts.
#[must_use]
pub fn label(counts: Counts) -> String {
    format!("Copied · {}", detail(counts))
}

/// Deterministic thousands grouping (`1204` → `"1,204"`).
///
/// Written out rather than taken from a locale formatter on purpose: the instrument voice must read
/// identically on every machine, and the label's pins must not drift with a system setting nobody
/// changed for this app's benefit.
#[must_use]
pub fn grouped(value: usize) -> String {
    let digits = value.to_string();
    let len = digits.len();
    let mut out = String::with_capacity(len + len.div_euclid(3));
    for (index, digit) in digits.chars().enumerate() {
        if index > 0 && (len - index).is_multiple_of(3) {
            out.push(',');
        }
        out.push(digit);
    }
    out
}

#[cfg(test)]
mod tests {
    use super::{Counts, counts, detail, grouped, label};

    #[test]
    fn a_shell_line_copy_is_one_line_and_its_trailing_newline_is_not_a_second() {
        assert_eq!(counts("foo\n"), Counts { chars: 4, lines: 1 });
        assert_eq!(counts("foo"), Counts { chars: 3, lines: 1 });
        assert_eq!(counts(""), Counts { chars: 0, lines: 1 });
    }

    /// Only one trailing newline is absorbed — the clipboard really does end with a blank line
    /// here, and a receipt that hid it would describe a different clipboard.
    #[test]
    fn a_second_trailing_newline_is_a_real_blank_line() {
        assert_eq!(counts("a\n\n").lines, 2);
        assert_eq!(counts("\n").lines, 1);
        assert_eq!(counts("a\nb\nc").lines, 3);
    }

    /// The chars figure is what Swift's `String.count` answers — grapheme CLUSTERS — because that
    /// is the number the label printed before it crossed, and a family emoji is one character
    /// to the person who selected it.
    #[test]
    fn characters_are_clusters_and_never_scalars_or_bytes() {
        assert_eq!(counts("é").chars, 1, "precomposed");
        assert_eq!(
            counts("e\u{301}").chars,
            1,
            "decomposed — one cluster, two scalars"
        );
        assert_eq!(
            counts("👨‍👩‍👧‍👦").chars,
            1,
            "one family, four scalars and three joiners"
        );
    }

    #[test]
    fn a_single_line_grab_speaks_characters_and_never_says_one_line() {
        assert_eq!(detail(counts("hello")), "5 characters");
        assert_eq!(detail(counts("x")), "1 character");
        assert_eq!(detail(counts("")), "0 characters");
    }

    #[test]
    fn a_multi_line_grab_speaks_lines() {
        assert_eq!(detail(counts("a\nb")), "2 lines");
        assert_eq!(label(counts("a\nb")), "Copied · 2 lines");
    }

    #[test]
    fn the_grouping_is_the_apps_own_and_not_the_machines() {
        assert_eq!(grouped(0), "0");
        assert_eq!(grouped(999), "999");
        assert_eq!(grouped(1_000), "1,000");
        assert_eq!(grouped(1_204), "1,204");
        assert_eq!(grouped(12_345), "12,345");
        assert_eq!(grouped(1_234_567), "1,234,567");
    }
}
